#**Linear Regression**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df = pd.read_csv('/content/Marketing_Data-Linear.csv')

In [ ]:
df.head()

youtube  facebook  newspaper  sales
0    84.72     19.20      48.96  12.60
1   351.48     33.96      51.84  25.68
2   135.48     20.88      46.32  14.28
3   116.64      1.80      36.00  11.52
4   318.72     24.00       0.36  20.88

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171 entries, 0 to 170
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   youtube    171 non-null    float64
 1   facebook   171 non-null    float64
 2   newspaper  171 non-null    float64
 3   sales      171 non-null    float64
dtypes: float64(4)
memory usage: 5.5 KB


In [ ]:
df.isnull().sum()

youtube      0
facebook     0
newspaper    0
sales        0
dtype: int64

In [ ]:
df.shape

(171, 4)

In [ ]:
#splitting into dependent and independent variable
x = df.iloc[:,:-1]
y = df['sales']

In [ ]:
x.head()

youtube  facebook  newspaper
0    84.72     19.20      48.96
1   351.48     33.96      51.84
2   135.48     20.88      46.32
3   116.64      1.80      36.00
4   318.72     24.00       0.36

In [ ]:
y.head()

0    12.60
1    25.68
2    14.28
3    11.52
4    20.88
Name: sales, dtype: float64

In [ ]:
x.shape, y.shape

((171, 3), (171,))

In [ ]:
#splitting into training and testing set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size = 0.33, random_state = 0)

In [ ]:
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
X_train = sc_x.fit_transform(X_train)
X_test  = sc_x.fit_transform(X_test)

In [ ]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
#train the model using the training sets
reg.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = reg.predict(X_test)

In [ ]:
from sklearn.metrics import r2_score
print("Accuracy" , r2_score(y_test , y_pred)*100,"%")

Accuracy 83.85155449906483 %


**After PCA**

In [ ]:
from sklearn.decomposition import PCA

the fit() method will allow us to get the parameters of the scaling function. The transform() method will transform the dataset to proceed with further data analysis steps. The fit_transform() method will determine the parameters and transform the dataset.

In [ ]:
pca = PCA(n_components=2)
X_train = pca.fit_transform(X_train,y_train)
X_test = pca.transform(X_test)

In [ ]:
reg = LinearRegression()
reg.fit(X_train,y_train)
y_pred = reg.predict(X_test)

In [ ]:
print("Accuracy" , r2_score(y_test , y_pred)*100,"%")

Accuracy 78.23039736600781 %


***After applying PCA to Regression model the acuraccy is decreased***

*   Before PCA - Accuracy 83.85 %
*   After PCA  - Accuracy 78.23 %




# **KNeighborsClassifier**

In [ ]:
kn = pd.read_csv('/content/Walmart_Store_sales-knn.csv')

In [ ]:
kn.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

In [ ]:
kn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


In [ ]:
kn.shape

(6435, 8)

In [ ]:
kn.isnull().sum()

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [ ]:
kn.drop(['Date'], axis = 1,inplace = True)

In [ ]:
kn = kn.loc[:, ["Holiday_Flag","Temperature","Fuel_Price","CPI","Unemployment","Weekly_Sales","Store"]]
kn.head()

Holiday_Flag  Temperature  Fuel_Price         CPI  Unemployment  \
0             0        42.31       2.572  211.096358         8.106   
1             1        38.51       2.548  211.242170         8.106   
2             0        39.93       2.514  211.289143         8.106   
3             0        46.63       2.561  211.319643         8.106   
4             0        46.50       2.625  211.350143         8.106   

   Weekly_Sales  Store  
0    1643690.90      1  
1    1641957.44      1  
2    1611968.17      1  
3    1409727.59      1  
4    1554806.68      1

In [ ]:
#defining dependent and independent variable
x= kn.iloc[:,:-1]
y = kn['Store']

In [ ]:
#Splitting the dataset into training and tescting set
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3, random_state=0)
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((4504, 6), (1931, 6), (4504,), (1931,))

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=11)
knn.fit(x_train,y_train)
y_pred= knn.predict(x_test)
y_pred[0:5]

array([18, 41, 37, 37, 25])

In [ ]:
#Test Model accuracy
from sklearn.metrics import accuracy_score

print("Accuracy", accuracy_score(y_test , y_pred)*100, "%")

Accuracy 21.957534955981355 %


**After PCA**

In [ ]:
from sklearn.decomposition import PCA
pc = PCA(n_components = 2)
x_train = pca.fit_transform(x_train,y_train)
x_test = pca.transform(x_test)

In [ ]:
knn = KNeighborsClassifier(10)
knn.fit(x_train,y_train)
y_pred = knn.predict(x_test)

In [ ]:
print("Accuracy-" ,accuracy_score(y_test , y_pred)*100,"%")

Accuracy- 21.957534955981355 %


***After applying PCA to KNN model the acuraccy is Same***

*   Before PCA - Accuracy 21.957 %
*   After PCA  - Accuracy 21.957 %

#**Logistic regression**

In [ ]:
df=pd.read_csv("/content/Amazon Sale Report.csv")
df.head()

<ipython-input-39-4ab493a2af69>:1: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("/content/Amazon Sale Report.csv")


index             Order ID      Date                        Status  \
0      0  405-8078784-5731545  04-30-22                     Cancelled   
1      1  171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   
2      2  404-0687676-7273146  04-30-22                       Shipped   
3      3  403-9615377-8133951  04-30-22                     Cancelled   
4      4  407-1069790-7240320  04-30-22                       Shipped   

  Fulfilment Sales Channel  ship-service-level    Style              SKU  \
0   Merchant      Amazon.in           Standard   SET389   SET389-KR-NP-S   
1   Merchant      Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL   
2     Amazon      Amazon.in          Expedited  JNE3371    JNE3371-KR-XL   
3   Merchant      Amazon.in           Standard    J0341       J0341-DR-L   
4     Amazon      Amazon.in          Expedited  JNE3671  JNE3671-TU-XXXL   

        Category  ... currency  Amount    ship-city   ship-state  \
0            Set  ...      INR  647.62       MUMBAI  MAHARASHTRA   
1          kurta  ...      INR  406.00    BENGALURU    KARNATAKA   
2          kurta  ...      INR  329.00  NAVI MUMBAI  MAHARASHTRA   
3  Western Dress  ...      INR  753.33   PUDUCHERRY   PUDUCHERRY   
4            Top  ...      INR  574.00      CHENNAI   TAMIL NADU   

  ship-postal-code  ship-country  \
0         400081.0            IN   
1         560085.0            IN   
2         410210.0            IN   
3         605008.0            IN   
4         600073.0            IN   

                                       promotion-ids    B2B  fulfilled-by  \
0                                                NaN  False     Easy Ship   
1  Amazon PLCC Free-Financing Universal Merchant ...  False     Easy Ship   
2       IN Core Free Shipping 2015/04/08 23-48-5-108   True           NaN   
3                                                NaN  False     Easy Ship   
4                                                NaN  False           NaN   

  Unnamed: 22  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 24 columns]

In [ ]:
df.drop(['index','Order ID','Date','Style','SKU','ASIN','ship-postal-code','promotion-ids','ship-city','B2B','Unnamed: 22'], axis=1, inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Status              128975 non-null  object 
 1   Fulfilment          128975 non-null  object 
 2   Sales Channel       128975 non-null  object 
 3   ship-service-level  128975 non-null  object 
 4   Category            128975 non-null  object 
 5   Size                128975 non-null  object 
 6   Courier Status      122103 non-null  object 
 7   Qty                 128975 non-null  int64  
 8   currency            121180 non-null  object 
 9   Amount              121180 non-null  float64
 10  ship-state          128942 non-null  object 
 11  ship-country        128942 non-null  object 
 12  fulfilled-by        39277 non-null   object 
dtypes: float64(1), int64(1), object(11)
memory usage: 12.8+ MB


In [ ]:
df.isnull().sum()

Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Category                  0
Size                      0
Courier Status         6872
Qty                       0
currency               7795
Amount                 7795
ship-state               33
ship-country             33
fulfilled-by          89698
dtype: int64

In [ ]:
df = df[(df['Status'] == 'Cancelled') | (df['Status'] == 'Shipped') | (df['Status'] == 'Shipped - Delivered to Buyer') | (df['Status'] == 'Shipped - Out for Delivery') | (df['Status'] == 'Pending') | (df['Status'] == 'Shipping')]
df["Status"] = df.Status.map({"Cancelled":0, "Shipped" : 1, "Shipped - Delivered to Buyer":2, "Shipped - Out for Delivery":3, "Pending":4, "Shipping":5})

In [ ]:
df.rename(columns = {'Courier Status': 'Courier_Status'}, inplace = True) 

In [ ]:
df = df[(df['Courier_Status'] == 'Shipped') | (df['Courier_Status'] == 'Unshipped')]
df['Courier_Status'] = df.Courier_Status.map({"Shipped":1, "Unshipped":0})

<ipython-input-46-183b5e54df9c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Courier_Status'] = df.Courier_Status.map({"Shipped":1, "Unshipped":0})


In [ ]:
df["Fulfilment"] = df.Fulfilment.map({"Merchant":0, "Amazon" : 1})

In [ ]:
df.rename(columns = {'Sales Channel ': 'Sales_Channel'}, inplace = True) 
df['Sales_Channel'] = df.Sales_Channel.map({"Amazon.in":1, "Non-Amazon":0})

In [ ]:
df.rename(columns = {'ship-service-level': 'ship_service_level'}, inplace = True) 
df['ship_service_level'] = df.ship_service_level.map({"Standard":1, "Expedited":0})

In [ ]:
df["Category"] = df.Category.map({"kurta":0, "Top":1, "Set":2, "Western Dress":3, "Ethnic Dress":4, "Saree":5, "Blouse":6, "Bottom":7, "Dupatta":8})

In [ ]:
df["Size"] = df.Size.map({"XS":0, "S":1, "M":2, "L":3, "XL":4, "XXL":5, "3XL":6, "4XL":7, "5XL":8, "6XL":9, "Free":10 })

In [ ]:
df['currency'] = df.currency.map({"INR":1, "nan":0})
df['currency'].fillna('0')

1         1.0
2         1.0
4         1.0
5         1.0
6         1.0
         ... 
128970    1.0
128971    1.0
128972    1.0
128973    1.0
128974    1.0
Name: currency, Length: 112802, dtype: object

In [ ]:
df['currency'].fillna(value = 0, inplace = True)
df['Amount'].fillna(value = 0, inplace = True)

In [ ]:
df.rename(columns= {"fulfilled-by" : "fulfilled_by"},inplace = True)
df['fulfilled_by'] = df['fulfilled_by'].fillna('Difficult Ship')
df["fulfilled_by"] = df.fulfilled_by.map({"Easy Ship":1, "Difficult Ship":0})

In [ ]:
df1 = df.loc[:,['Courier_Status','Status','Fulfilment','Sales_Channel','ship_service_level','Category','Size','Qty','ship_country','fulfilled_by','currency','Amount']]

In [ ]:
df1.isnull().sum()

Courier_Status             0
Status                     0
Fulfilment                 0
Sales_Channel              0
ship_service_level         0
Category                   0
Size                       0
Qty                        0
ship_country               0
fulfilled_by          112802
currency                   0
Amount                     0
dtype: int64

In [ ]:
df1.drop(['fulfilled_by'],inplace = True, axis = 1)

In [ ]:
df1.head()

Courier_Status  Status  Fulfilment  Sales_Channel  ship_service_level  \
1               1       2           0              1                   1   
2               1       1           1              1                   0   
4               1       1           1              1                   0   
5               1       1           1              1                   0   
6               1       1           1              1                   0   

   Category  Size  Qty  ship_country  currency  Amount  
1         0     6    1             1       1.0   406.0  
2         0     4    1             1       1.0   329.0  
4         1     6    1             1       1.0   574.0  
5         2     4    1             1       1.0   824.0  
6         2     3    1             1       1.0   653.0

In [ ]:
#Split dataset into independent and dependent features
X = df1.iloc[:,2:]
y = df1.iloc[:,1]

In [ ]:
#splitting the dataset into the training and testing set
from sklearn.model_selection import train_test_split
X_Train, X_Test, y_Train, y_Test = train_test_split(X,y,test_size = 0.25, random_state = 0)
X_Train.shape, y_Train.shape, X_Test.shape, y_Test.shape

((84601, 9), (84601,), (28201, 9), (28201,))

In [ ]:
#Scaling the Data
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_Train = sc_X.fit_transform(X_Train)
X_Test = sc_X.transform(X_Test)

In [ ]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_Train,y_Train)

/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=0)

In [ ]:
y_Pred = classifier.predict(X_Test)

In [ ]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(y_Test, y_Pred)
print("Accuracy", acc*100, "%")

Accuracy 94.39381582213396 %


**After PCA**

In [ ]:
pca = PCA(n_components=2)
X_train = pca.fit_transform(X_Train,y_Train)
X_test = pca.transform(X_Test)

In [ ]:
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_Train, y_Train)

/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=0)

In [ ]:
Pred = classifier.predict(X_Test)

In [ ]:
#Accuracy

from sklearn.metrics import accuracy_score
acc = accuracy_score(y_Test, Pred)
print("Accuracy", acc*100, "%")

Accuracy 94.39381582213396 %


***After applying PCA to Logistic Regression model the acuraccy is Same***

*   Before PCA - Accuracy 94.39 %
*   After PCA  - Accuracy 94.39 %